compiler/codegen/cg-expr.cpp

```diff
@@ -4720,8 +4720,37 @@ static GenRet codegenGPUKernelLaunch(CallExpr* call, bool is3d) {
   std::vector<GenRet> args;
   int curArg = 1;
   for_actuals(actual, call) {
-    if (curArg > nNonKernelParamArgs) {
-      args.push_back(codegenAddrOf(actual));
+    Symbol* actualSym = toSymExpr(actual)->symbol();
+    if (curArg == 1) {
+      if (FnSymbol* fn = toFnSymbol(actualSym)) {
```

Comment:
I assume this has to do with this bullet in your PR description:

> Allows passing function symbols to the primitive. This enables having multiple kernels/loops to be run on the GPU.

I guess I get line 4725: you check to see if it's a function symbol and if so you return a c string to its name. I assume this name ends up being unique-ified if we have multiple outlined kernels.

What does `isImmediate()` return?  I guess the idea behind this branch is if what's being passed in is already a C string you just codegen it and forward it?  Maybe we should assert that this argument type is something that codegen's to a CStringSymbol?


Assuming my understanding is correct could you update the comment above (starting on line 4699) to indicate the first parameter could either be the function name or a function symbol:

```C++
  // The primitive's arguments are
  //   - function name
  //   - grid size (1 arg or 3 args)
  //   - block size (1 arg or 3 args)
  //   - any number of arguments to be passed to the kernel
 ...
```

Reply:
Your assumptions are all correct.

> What does isImmediate() return? I guess the idea behind this branch is if what's being passed in is already a C string you just codegen it and forward it? Maybe we should assert that this argument type is something that codegen's to a CStringSymbol?

Yep, I'll add a check to make sure that the immediate is in fact a c_string.

I'll update the comment, too.